### PR TITLE
Align qesap regression LTSS to HanaSR

### PR DESCRIPTION
**Pull request by mpagot**
Change the qesap regression implementation of LTSS registration to be more aligned to the one implemented for HanaSR. Use more similar variables, use functions from the registration library. Allow LTSS to be enabled in all conf.yaml.

Related ticket: https://jira.suse.com/browse/TEAM-9839

Improved version of https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20624

# Verification run:
https://openqaworker15.qa.suse.cz/tests/overview?distri=sle&version=12-SP5&build=mpagot_VR&groupid=40

### Azure
http://openqaworker15.qa.suse.cz/tests/306482 LTSS-ES register.yaml command line is correctly composed in http://openqaworker15.qa.suse.cz/tests/306482/file/configure-qesap_azure.yaml `-e use_suseconnect=true -e sles_modules='[{"key":"SLES-LTSS-Extended-Security/12.5/x86_64","value":"********"}]'`  and registration is ok in http://openqaworker15.qa.suse.cz/tests/306482/logfile?filename=deploy-ansible.registration.log.txt  :green_circle: 

http://openqaworker15.qa.suse.cz/tests/306483 no LTSS :green_circle: 

### Aws
http://openqaworker15.qa.suse.cz/tests/306486 no LTSS :green_circle: 
http://openqaworker15.qa.suse.cz/tests/306487 LTSS :green_circle:  

### Gcp
http://openqaworker15.qa.suse.cz/tests/306490 no  LTSS :green_circle:  registration is fine, no LTSS module, failure is later in the cluster status (native fencing)
http://openqaworker15.qa.suse.cz/tests/306491  LTSS :green_circle:  registration is fine, LTSS module enabled, failure is later in the cluster status (native fencing)

### Azure
http://openqaworker15.qa.suse.cz/tests/306501 no LTSS :green_circle: 
http://openqaworker15.qa.suse.cz/tests/306500 :red_circle: failure in LTSS registration, known issue, not test issue

### Aws
http://openqaworker15.qa.suse.cz/tests/306498  no LTSS :green_circle: 
http://openqaworker15.qa.suse.cz/tests/306499   LTSS  registration is fine but later `zypper ref` timeout
https://openqaworker15.qa.suse.cz/tests/306567   LTSS  registration is fine but later `zypper ref` timeout

### Gcp
http://openqaworker15.qa.suse.cz/tests/306569 no LTSS :green_circle:  registration is fine, LTSS module enabled, failure is later in the cluster status (native fencing)
http://openqaworker15.qa.suse.cz/tests/306570 LTSS :red_circle: fails in `zypper ref`

### CODE REVIEW

File: tests/sles4sap/qesapdeployment/configure.pm [hunk 1]
```diff
@@ -47,7 +48,26 @@ sub run {
...
+    my $reg_args;
+    $reg_args = "-e use_suseconnect=true " if (get_var('QESAPDEPLOY_USE_SUSECONNECT'));
+    my @addons = grep { defined $_ && $_ } split(/,/, get_var('SCC_ADDONS'));
+    # This implementation has a known limitation
+    # if SCC_ADDONS has two or more elements (like "ltss,ltss_es")
+    # only the first one will be added to the playbook argument.
+    foreach my $addon (@addons) {
+        my $name;
+        # Keep the code simple by only support ltss addons,
+        # it simplify version calculation.
```
Comment (alvarocarvajald) [inline tests/sles4sap/qesapdeployment/configure.pm:61]: s/simplify/simplifies/

File: tests/sles4sap/qesapdeployment/configure.pm [hunk 2]
```diff
@@ -47,7 +48,26 @@ sub run {
     $variables{SLES4SAP_PUBSSHKEY} = get_ssh_private_key_path() . '.pub';
     $variables{REGISTRATION_PLAYBOOK} = get_var('QESAPDEPLOY_REGISTRATION_PLAYBOOK', 'registration');
     $variables{REGISTRATION_PLAYBOOK} =~ s/\.yaml$//;
-    $variables{SUSECONNECT} = get_var('QESAPDEPLOY_USE_SUSECONNECT', 'false');
+
+    my $reg_args;
+    $reg_args = "-e use_suseconnect=true " if (get_var('QESAPDEPLOY_USE_SUSECONNECT'));
+    my @addons = grep { defined $_ && $_ } split(/,/, get_var('SCC_ADDONS'));
```
Comment (alvarocarvajald) [inline tests/sles4sap/qesapdeployment/configure.pm:54]:
```suggestion
    my @addons = split(/,/, get_var('SCC_ADDONS', ''));
```
No need for the `grep`. Splitting an empty string should produce an empty list.
```
$ perl
use strict;
use warnings;
my $t = '';
my @l = split(/,/, $t);
use Data::Dumper;
say Dumper \@l;
$VAR1 = [];

```

File: tests/sles4sap/qesapdeployment/configure.pm [hunk 3]
```diff
@@ -47,7 +48,26 @@ sub run {
...
+    foreach my $addon (@addons) {
+        my $name;
+        # Keep the code simple by only support ltss addons,
+        # it simplify version calculation.
+        $name = get_addon_fullname($addon) if ($addon =~ 'ltss');
+        if ($name) {
+            record_info($name, "Register '$name' with code '$ADDONS_REGCODE{$name}'");
+            $reg_args .= qesap_ansible_reg_module(reg => join(',', join('/', $name, scc_version(), 'x86_64'), $ADDONS_REGCODE{$name}));
+            # Break not to pack on the playbook command line more than one "-e sles_module"
+            break;
```
Comment (alvarocarvajald) [inline tests/sles4sap/qesapdeployment/configure.pm:67]:
```suggestion
            last;
```

File: lib/qesapdeployment.pm [hunk 4]
```diff
@@ -1133,6 +1134,37 @@ sub qesap_ansible_fetch_file {
...
 }
 
+=head3 qesap_ansible_reg_module
+
+    Compose the ansible-playbook argument for the registration.yaml playbook,
+    about an additional module registration
+
+    -e sles_modules='[{"key":"SLES-LTSS-Extended-Security/12.5/x86_64","value":"*******"}]'
+
+    Known limitation is that registration.yaml support multiple modules to be registered,
```
Comment (alvarocarvajald) [inline lib/qesapdeployment.pm:1144]: s/support/supports/

File: lib/qesapdeployment.pm [hunk 5]
```diff
@@ -1133,6 +1134,37 @@ sub qesap_ansible_fetch_file {
...
+    this code only support one.
```
Comment (alvarocarvajald) [inline lib/qesapdeployment.pm:1145]: s/support/supports/

File: lib/qesapdeployment.pm [hunk 6]
```diff
@@ -1133,6 +1134,37 @@ sub qesap_ansible_fetch_file {
...
+                Second string has to be valid registration code for the particular extension.
+
+=back
+=cut
+
+sub qesap_ansible_reg_module {
+    my (%args) = @_;
+    croak 'Missing mandatory "reg" argument' unless $args{reg};
+    my @reg_args = split(/,/, $args{reg});
+    die "Missing reg_code for '$reg_args[0]'" if scalar @reg_args != 2;
```
Comment (alvarocarvajald) [inline lib/qesapdeployment.pm:1162]:
```suggestion
    die "Missing reg_code for '$reg_args[0]'" if (@reg_args != 2);
```
`scalar` keyword is not needed. `!= 2` will make Perl treat `@reg_args` as an scalar.

File: tests/sles4sap/qesapdeployment/configure.pm [hunk 7]
```diff
@@ -47,7 +48,26 @@ sub run {
...
+    # This implementation has a known limitation
+    # if SCC_ADDONS has two or more elements (like "ltss,ltss_es")
+    # only the first one will be added to the playbook argument.
+    foreach my $addon (@addons) {
+        my $name;
+        # Keep the code simple by only support ltss addons,
+        # it simplify version calculation.
+        $name = get_addon_fullname($addon) if ($addon =~ 'ltss');
+        if ($name) {
+            record_info($name, "Register '$name' with code '$ADDONS_REGCODE{$name}'");
```
Comment (alvarocarvajald) [inline tests/sles4sap/qesapdeployment/configure.pm:64]: This is for debug purposes, right? I think it's better not to print the regcode.

File: lib/qesapdeployment.pm [hunk 8]
```diff
@@ -1133,6 +1134,37 @@ sub qesap_ansible_fetch_file {
...
+                Second string has to be valid registration code for the particular extension.
+
+=back
+=cut
+
+sub qesap_ansible_reg_module {
+    my (%args) = @_;
+    croak 'Missing mandatory "reg" argument' unless $args{reg};
+    my @reg_args = split(/,/, $args{reg});
+    die "Missing reg_code for '$reg_args[0]'" if (@reg_args != 2);
```
Comment (Amrysliu) [inline lib/qesapdeployment.pm:1162]: `scalar` doesn't need here?

**Review by Amrysliu — APPROVED** [review]
LGTM

File: lib/sles4sap_publiccloud.pm [hunk 9]
```diff
@@ -941,13 +934,7 @@ sub create_playbook_section_list {
...
         push @reg_args, "-e reg_code=$args{scc_code} -e email_address=''";
         push @reg_args, '-e use_suseconnect=true' if ($args{registration} eq 'suseconnect');
-        if ($args{ltss}) {
-            my @ltss_args = split(/,/, $args{ltss});
-            die "Missing reg_code for '$ltss_args[0]'" if scalar @ltss_args != 2;
-            push @reg_args, "-e sles_modules='[{" .
-              "\"key\":\"$ltss_args[0]\"," .
-              "\"value\":\"$ltss_args[1]\"}]'";
-        }
+        push @reg_args, qesap_ansible_reg_module(reg => $args{ltss}) if ($args{ltss});
```
Comment (BillAnastasiadis) [inline lib/sles4sap_publiccloud.pm:937]: I see that `args{ltss}` is still used, shouldn't it be kept in the docstring?

**Review by alvarocarvajald — APPROVED** [review]
LGTM

File: lib/sles4sap_publiccloud.pm (hunk 9 above)
Comment (mpagot) [reply lib/sles4sap_publiccloud.pm:937]: it is, few line above https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20796/files#diff-4b7291e584e1d0b487f685a7b45efc74dcd804ee2236acf7a866003d0f4fb87aR889

**Pull request merged**